core/src/main/java/org/apache/iceberg/SerializableTable.java

```diff
@@ -168,23 +172,27 @@ public Schema schema() {
 
   @Override
   public PartitionSpec spec() {
-    if (lazySpec == null) {
+    return specs().get(defaultSpecId);
+  }
+
+  @Override
+  public Map<Integer, PartitionSpec> specs() {
+    if (lazySpecs == null) {
       synchronized (this) {
-        if (lazySpec == null && lazyTable == null) {
+        if (lazySpecs == null && lazyTable == null) {
           // prefer parsing JSON as opposed to loading the metadata
-          this.lazySpec = PartitionSpecParser.fromJson(schema(), specAsJson);
-        } else if (lazySpec == null) {
-          this.lazySpec = lazyTable.spec();
+          Map<Integer, PartitionSpec> parsedSpecs = Maps.newHashMap();
```

Comment:
Can we optimize by newHashMapWithExpectedSize?

Reply:
Fixed.